File: sdks/ts/memory/src/ingest/safety.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Prompt injection safety scanner for the ingest pipeline. Uses
 * @stackone/defender for ML-based detection (Tier 2: MiniLM classifier).
 *
 * The scanner is OPTIONAL: if the defender package is unavailable or the
 * model fails to load, ingestion continues with a warning. Flagged
 * documents are never blocked — they receive metadata annotations so
 * downstream consumers can handle them appropriately.
 */

import type { Logger } from '../llm/types.js'

/** Confidence threshold below which content is considered safe. */
const INJECTION_CONFIDENCE_THRESHOLD = 0.5

/** Tag used to wrap ingested content for isolation. */
const ISOLATION_TAG = 'ingested-document'

/** Characters that should be stripped during preprocessing (zero-width). */
const ZERO_WIDTH_CHARS_RE =
  /(?:[\u200B\u200C\u200E\u200F\u2000\u2001\u2060\u2061\u2062\u2063\u2064\uFEFF\u00AD]|\u200D)/g

/** Pattern to detect closing isolation tags that must be escaped in content. */
const CLOSING_TAG_RE = /<\/ingested-document>/gi

/** Escaped replacement for closing tags found in content. */
const CLOSING_TAG_ESCAPED = '&lt;/ingested-document&gt;'

export type SafetyScanResult = {
  readonly injectionDetected: boolean
  readonly confidence: number
  readonly detections: readonly string[]
}

export type IsolatedContent = {
  readonly content: string
  readonly source: string
  readonly hash: string
}

export type SafetyMetadata = {
  readonly injection_risk: boolean
  readonly injection_confidence: number
}

export type SafetyScannerConfig = {
  readonly logger: Logger
  readonly confidenceThreshold?: number
}

/**
 * Normalise text for consistent scanning: NFKC normalisation and removal
 * of zero-width characters that could evade pattern detection.
 *
 * Time: O(n) where n = text length.
 * Space: O(n) for the normalised copy.
 */
export const preprocessText = (text: string): string => {
  const normalised = text.normalize('NFKC')
  return normalised.replace(ZERO_WIDTH_CHARS_RE, '')
}

/**
 * Wrap content in isolation delimiters to prevent prompt injection
 * breakout. Any existing closing tags in the content are escaped.
 *
 * Time: O(n) where n = content length.
 * Space: O(n) for the wrapped string.
 */
export const wrapInIsolation = ({ content, source, hash }: IsolatedContent): string => {
  const escapedContent = content.replace(CLOSING_TAG_RE, CLOSING_TAG_ESCAPED)
  return `<${ISOLATION_TAG} source="${escapeAttr(source)}" hash="${hash}">${escapedContent}</${ISOLATION_TAG}>`
}

/**
 * Escape XML attribute characters to prevent attribute injection.
 */
const escapeAttr = (value: string): string =>
  value.replace(/&/g, '&amp;').replace(/"/g, '&quot;').replace(/</g, '&lt;').replace(/>/g, '&gt;')

/**
 * Create a safety scanner that uses @stackone/defender for ML-based
 * prompt injection detection. The scanner loads lazily: the ONNX model
 * is loaded on first scan. If the model is unavailable, the scanner
 * degrades gracefully and logs a warning.
 *
 * Returns undefined when the defender package cannot be imported.
 */
export const createSafetyScanner = (
  config: SafetyScannerConfig,
): { scan: (text: string, signal?: AbortSignal) => Promise<SafetyScanResult> } | undefined => {
  const { logger } = config
  const threshold = config.confidenceThreshold ?? INJECTION_CONFIDENCE_THRESHOLD

  let defenderInstance: DefenderLike | undefined
  let loadAttempted = false
  let loadFailed = false

  const ensureDefender = async (): Promise<DefenderLike | undefined> => {
    if (defenderInstance !== undefined) return defenderInstance
    if (loadFailed) return undefined
    if (loadAttempted) return undefined

    loadAttempted = true
    try {
      const mod = await import('@stackone/defender')
      defenderInstance = mod.createPromptDefense({
        enableTier1: true,
        enableTier2: true,
        blockHighRisk: false,
      })
      await defenderInstance.warmupTier2()
      return defenderInstance
    } catch (err: unknown) {
      loadFailed = true
      logger.warn('safety scanner: defender unavailable, continuing without ML detection', {
        error: String(err),
      })
      return undefined
    }
  }

  const scan = async (text: string, _signal?: AbortSignal): Promise<SafetyScanResult> => {
    const preprocessed = preprocessText(text)
    const defender = await ensureDefender()

    if (defender === undefined) {
      return { injectionDetected: false, confidence: 0, detections: [] }
    }

    try {
      const result = await defender.defendToolResult(preprocessed, 'ingest_document')
      const confidence = result.tier2Score ?? 0
      const injectionDetected = confidence >= threshold
      const detections: string[] = [...result.detections]

      if (injectionDetected) {
        logger.info('safety scanner: injection detected', {
          confidence,
          detections,
          riskLevel: result.riskLevel,
        })
      }

      return { injectionDetected, confidence, detections }
    } catch (err: unknown) {
      logger.warn('safety scanner: scan failed, treating as safe', {
        error: String(err),
      })
      return { injectionDetected: false, confidence: 0, detections: [] }
    }
  }

  // Verify import is possible synchronously by attempting a dynamic import
  // We return the scanner optimistically; the actual load happens on first scan
  return { scan }
}

/**
 * Build metadata fields for a flagged document. Returns an empty object
 * when the scan did not detect injection, so spreading into metadata is
 * always safe.
 */
export const buildSafetyMetadata = (scanResult: SafetyScanResult): Partial<SafetyMetadata> => {
  if (!scanResult.injectionDetected) return {}
  return {
    injection_risk: true,
    injection_confidence: scanResult.confidence,
  }
}

/**
 * Minimal interface matching the subset of PromptDefense we consume.
 * Avoids coupling to the full @stackone/defender type surface.
 */
type DefenderLike = {
  warmupTier2(): Promise<void>
  defendToolResult(
    value: unknown,
    toolName: string,
  ): Promise<{
    allowed: boolean
    riskLevel: string
    detections: string[]
    tier2Score?: number
  }>
}

File: go/ingest/safety.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"regexp"
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

// injectionConfidenceThreshold is the score above which content is
// flagged as a potential prompt injection. The ML model returns scores
// in [0.0, 1.0]; 0.5 balances precision and recall.
const injectionConfidenceThreshold = 0.5

// isolationTag is the XML element name used to wrap ingested content,
// providing structural isolation from the host prompt.
const isolationTag = "ingested-document"

// closingTagPattern matches the isolation closing tag in any casing so
// it can be escaped to prevent content breakout.
var closingTagPattern = regexp.MustCompile(`(?i)</ingested-document>`)

// ScanResult holds the output of a prompt injection scan.
type ScanResult struct {
	InjectionDetected bool
	Confidence        float64
	Detections        []string
}

// SafetyMetadata contains the metadata fields added to flagged
// documents. Callers merge these into the document's metadata map.
type SafetyMetadata struct {
	InjectionRisk       bool
	InjectionConfidence float64
}

// Scanner is the interface for ML-based prompt injection detection.
// Implementations wrap specific model runtimes (ONNX via hugot,
// remote API, etc.). The interface allows graceful degradation: when
// no scanner is available the pipeline continues without ML scoring.
type Scanner interface {
	// Scan analyses preprocessed text for prompt injection indicators.
	// Implementations must respect context cancellation.
	Scan(ctx context.Context, text string) (ScanResult, error)

	// Available reports whether the underlying model is loaded and
	// ready to serve predictions.
	Available() bool
}

// ScannerConfig holds construction parameters for a Scanner.
type ScannerConfig struct {
	// ModelPath is the filesystem path to the ONNX model file.
	// Required for hugot-backed implementations.
	ModelPath string

	// ConfidenceThreshold overrides the default injection threshold.
	// Zero means use the package default (0.5).
	ConfidenceThreshold float64
}

// noopScanner is the fallback scanner returned when no ML model is
// available. It never flags content and always reports unavailable.
type noopScanner struct{}

func (noopScanner) Scan(_ context.Context, _ string) (ScanResult, error) {
	return ScanResult{InjectionDetected: false, Confidence: 0, Detections: nil}, nil
}

func (noopScanner) Available() bool { return false }

// NewNoopScanner returns a Scanner that performs no ML inference. Use
// this as the default when the ONNX model is not available.
func NewNoopScanner() Scanner {
	return noopScanner{}
}

// PreprocessText normalises content for reliable scanning. Applies
// Unicode NFKC normalisation and strips zero-width characters commonly
// used to evade text-based detection.
//
// Time: O(n) where n = number of runes.
// Space: O(n) for the normalised copy.
func PreprocessText(text string) string {
	normalised := norm.NFKC.String(text)
	var b strings.Builder
	b.Grow(len(normalised))
	for _, r := range normalised {
		if isZeroWidthChar(r) {
			continue
		}
		b.WriteRune(r)
	}
	return b.String()
}

// isZeroWidthChar reports whether a rune is a zero-width or invisible
// formatting character used for evasion.
func isZeroWidthChar(r rune) bool {
	switch r {
	case
		'​', // zero-width space
		'‌', // zero-width non-joiner
		'‍', // zero-width joiner
		'‎', // left-to-right mark
		'‏', // right-to-left mark
		'⁠', // word joiner
		'⁡', // function application
		'⁢', // invisible times
		'⁣', // invisible separator
		'⁤', // invisible plus
		'\ufeff', // zero-width no-break space (BOM)
		'­', // soft hyphen
		' ', // en quad (thin space, often abused)
		' ': // em quad
		return true
	}
	// Catch any remaining characters in the General_Category=Format class
	// that are not common whitespace.
	return unicode.Is(unicode.Cf, r) && !isCommonWhitespace(r)
}

// isCommonWhitespace identifies whitespace characters that should be
// preserved (space, tab, newline, carriage return).
func isCommonWhitespace(r rune) bool {
	switch r {
	case ' ', '\t', '\n', '\r':
		return true
	}
	return false
}

// WrapInIsolation encloses content in XML isolation delimiters. Any
// occurrences of the closing tag within the content are escaped to
// prevent delimiter breakout.
//
// Time: O(n) where n = content length.
// Space: O(n) for the wrapped string.
func WrapInIsolation(content, source, hash string) string {
	escapedContent := closingTagPattern.ReplaceAllString(content, "&lt;/ingested-document&gt;")
	escapedSource := escapeAttr(source)
	return fmt.Sprintf(
		"<%s source=%q hash=%q>%s</%s>",
		isolationTag, escapedSource, hash, escapedContent, isolationTag,
	)
}

// escapeAttr escapes XML special characters in attribute values.
func escapeAttr(value string) string {
	value = strings.ReplaceAll(value, "&", "&amp;")
	value = strings.ReplaceAll(value, "\"", "&quot;")
	value = strings.ReplaceAll(value, "<", "&lt;")
	value = strings.ReplaceAll(value, ">", "&gt;")
	return value
}

// ContentHash computes a SHA-256 hash of the content, returned as a
// hex string. Used for the isolation delimiter hash attribute.
func ContentHash(content []byte) string {
	sum := sha256.Sum256(content)
	return hex.EncodeToString(sum[:])
}

// BuildSafetyMetadata returns metadata fields for a flagged document.
// Returns nil when no injection was detected so callers can skip the
// merge.
func BuildSafetyMetadata(result ScanResult) *SafetyMetadata {
	if !result.InjectionDetected {
		return nil
	}
	return &SafetyMetadata{
		InjectionRisk:       true,
		InjectionConfidence: result.Confidence,
	}
}

// ScanAndWrap is the primary integration point: preprocess the content,
// run the scanner, wrap in isolation delimiters, and return the wrapped
// content plus any safety metadata.
func ScanAndWrap(
	ctx context.Context,
	scanner Scanner,
	content string,
	source string,
) (wrappedContent string, metadata *SafetyMetadata, err error) {
	preprocessed := PreprocessText(content)
	hash := ContentHash([]byte(preprocessed))

	var scanResult ScanResult
	if scanner.Available() {
		var scanErr error
		scanResult, scanErr = scanner.Scan(ctx, preprocessed)
		if scanErr != nil {
			// Graceful degradation: log via caller, do not block ingestion
			scanResult = ScanResult{InjectionDetected: false, Confidence: 0}
		}
	}

	wrapped := WrapInIsolation(preprocessed, source, hash)
	meta := BuildSafetyMetadata(scanResult)
	return wrapped, meta, nil
}
